addons/confirma/src/helpers/AssertionMessageGenerator.cs

```diff
@@ -0,0 +1,39 @@
+using System.Globalization;
+using Confirma.Formatters;
+
+namespace Confirma.Helpers;
+
+public class AssertionMessageGenerator
+{
+    private readonly string _format;
+    private readonly string _assertion;
+    private readonly object? _expected;
+    private readonly object? _actual;
+    private readonly Formatter _formatter;
+
+    public AssertionMessageGenerator(
+        string format,
+        string assertion,
+        Formatter formatter,
+        object? expected,
+        object? actual
+    )
+    {
+        _format = format;
+        _assertion = assertion;
+        _formatter = formatter;
+        _expected = expected;
+        _actual = actual;
+    }
+
+    public string GenerateMessage()
+    {
+        return string.Format(
+            CultureInfo.InvariantCulture,
+            "Assertion {0} failed: " + _format,
+            _assertion,
+            _expected ?? _formatter.Format(_expected!),
+            _actual ?? _formatter.Format(_actual!)
```

Comment:
Should it format itself only if `_expected` or `_actual` is `null`?

Reply:
if not, that fix will help with failing `ConfirmActionTest` tests 
```suggestion
            _formatter.Format(_expected!),
            _formatter.Format(_actual!)
```

Reply:
Unfortunately, this fix won't do because assertion messages can be constructed in 4 ways:

1. Actual & expected are present: e.q. "Expected {actual}  to be {expected}.",
2. Only actual is present: e.q. "Expected {actual} to be null.",
3. Only expected is present: e.q. "Expected to be {expected}.",
4. Nothing is present: e.q. "Expected non-null value.".

This fix would work only for the first one. Tomorrow I'll look for a solution because current state as you mentioned breaks formatting in `ConfirmActionTest`. For some reason, nullable values break `NumericFormatter` even when they are not null.